### PR TITLE
Adding reproducibility tests for point3d initializers

### PR DESCRIPTION
**Pull request by ayushbaid**
Adding the reproducibility tests, with the test being flaky.

The test for no-ransac configuration fails occasionally. Will require digging into the GTSAM code.

### CODE REVIEW

**Comment by ayushbaid** [comment]
> Thanks for working on this, Ayush. Was curious why the threshold affects the reproducibility.

The threshold matters because we were using some random noise in the input measurements. As a result, the threshold turned out to be too strict.

File: tests/repro_tests/data_association/test_point3dinitializer_reproducibility.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    TriangulationSamplingMode,
+    TriangulationExitCode,
+)
+from tests.repro_tests.test_repro_base import ReproducibilityTestBase
+
+# Load ground truth data with 3 images and 7 tracks.
+GTSAM_EXAMPLE_FILE: str = "dubrovnik-3-7-pre"
+GROUND_TRUTH_DATA: GtsfmData = io_utils.read_bal(gtsam.findExampleDataFile(GTSAM_EXAMPLE_FILE))
+
+MEASUREMENTS_NOISE_STANDARD_DEVIATION = 1  # in pixels
```
Comment (johnwlambert) [inline tests/repro_tests/data_association/test_point3dinitializer_reproducibility.py:28]: looks good. `MEASUREMENTS_NOISE_STD_DEV` would be a little more readable imho

File: tests/repro_tests/data_association/test_point3dinitializer_reproducibility.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+    def run_once(self) -> TRIANGULATION_RESULT_TYPE:
+        return self._point3d_initializer.triangulate(self._track_2d)
+
+    def assert_results(self, results_a: TRIANGULATION_RESULT_TYPE, results_b: TRIANGULATION_RESULT_TYPE) -> None:
+        self.__assert_track3d(results_a[0], results_b[0])
+        self.assertAlmostEqual(results_a[1], results_b[1])
+        self.assertEqual(results_a[2], results_b[2])
+
+    def __assert_track3d(self, track3d_a: Optional[SfmTrack], track3d_b: Optional[SfmTrack]) -> None:
```
Comment (johnwlambert) [inline tests/repro_tests/data_association/test_point3dinitializer_reproducibility.py:68]: the fn name sounds a bit like we are asserting this is a track. i think we also want to assert they're equal tracks, right?
would `__assert_track3d_equal()` make sense here?

Comment (johnwlambert) [reply tests/repro_tests/data_association/test_point3dinitializer_reproducibility.py:68]: or should we add this on the GTSAM side as a .equals() method?

File: tests/repro_tests/data_association/test_point3dinitializer_reproducibility.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        triangulation_options: TriangulationOptions = TriangulationOptions(
+            reproj_error_threshold=TRIANGULATION_REPROJ_ERROR_THRESHOLD,
+            mode=TriangulationSamplingMode.RANSAC_SAMPLE_UNIFORM,
+        ),
+    ) -> None:
+        super().setUp(triangulation_options)
+
+        # replace one measurement with an outlier
+        measurements_with_one_outlier = self._track_2d.measurements
+        measurements_with_one_outlier[2] = SfmMeasurement(measurements_with_one_outlier[2].i, np.array([50, 30]))
```
Comment (johnwlambert) [inline tests/repro_tests/data_association/test_point3dinitializer_reproducibility.py:95]: maybe an inline comment stating what the correct value was before?

**Comment by johnwlambert** [comment]
thanks for the PR, @ayushbaid.

**Comment by johnwlambert** [comment]
> > Thanks for working on this, Ayush. Was curious why the threshold affects the reproducibility.
> 
> The threshold matters because we were using some random noise in the input measurements. As a result, the threshold turned out to be too strict.

Hmm not sure I follow still. Could you give a brief example here of what you mean?

**Pull request closed without merging**